### PR TITLE
Do not read files from disk

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

**Review by MBkkt — APPROVED** [review]

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/utils/directory_utils.cpp [hunk 1]
```diff
@@ -65,17 +65,19 @@ bool RemoveAllUnreferenced(directory& dir) {
 
 TrackingDirectory::TrackingDirectory(directory& impl,
                                      bool track_open /*= false*/) noexcept
-  : impl_(impl), track_open_(track_open) {}
+  : impl_{impl},
+    on_close_{[this](size_t size) noexcept { this->files_size_ += size; }},
```
Comment (MBkkt) [inline core/utils/directory_utils.cpp:69]: NIT
```suggestion
    on_close_{[&](size_t size) noexcept { files_size_ += size; }},
```

File: core/utils/directory_utils.cpp [hunk 2]
```diff
@@ -65,17 +65,19 @@ bool RemoveAllUnreferenced(directory& dir) {
...
+    return nullptr;
   }
 
-  auto result = impl_.create(name);
-
-  if (result) {
+  if (auto result = impl_.create(name); result) {
+    result->SetCallback(&on_close_);
     return result;
   }
```
Comment (MBkkt) [inline core/utils/directory_utils.cpp:83]:
```cpp
index_output::ptr TrackingDirectory::create(std::string_view name) noexcept try {
  files_.emplace(name);
  if (auto result = impl_.create(name); result) {
    result->SetCallback(&on_close_);
    return result;
  }
  files_.erase(name);
  return nullptr;
} catch (...) {
  return nullptr;
}
```

File: core/utils/directory_utils.cpp [hunk 3]
```diff
@@ -133,16 +135,20 @@ bool TrackingDirectory::rename(std::string_view src,
...
 
-void TrackingDirectory::clear_tracked() noexcept { files_.clear(); }
+void TrackingDirectory::ClearTracked() noexcept {
+  files_size_ = 0;
+  files_.clear();
+}
 
-std::vector<std::string> TrackingDirectory::flush_tracked() {
+std::vector<std::string> TrackingDirectory::FlushTracked(size_t& files_size) {
   std::vector<std::string> files(files_.size());
```
Comment (MBkkt) [inline core/utils/directory_utils.cpp:144]:
```suggestion
void TrackingDirectory::FlushTracked(std::vector<std::string>& files, size_t& files_size) {
  files.resize(files_.size());
```

File: core/index/index_writer.hpp [hunk 4]
```diff
@@ -939,6 +939,7 @@ class IndexWriter : private util::noncopyable {
   // Start transaction
   bool Start(ProgressReportCallback const& progress = nullptr);
   void StartImpl(FlushContextPtr&& ctx, DirectoryMeta&& to_commit,
+                 std::vector<SegmentReader> readers,
                  std::vector<std::string_view> files_to_sync);
```
Comment (MBkkt) [inline core/index/index_writer.hpp:943]:
```suggestion
                 std::vector<SegmentReader>&& readers,
                 std::vector<std::string_view>&& files_to_sync);
```

File: core/index/index_writer.cpp [hunk 5]
```diff
@@ -320,17 +320,31 @@ size_t WriteDocumentMask(directory& dir, SegmentMeta& meta,
...
+    size_t size;
+    if (!dir.length(size, file)) {
+      throw io_error{
+        absl::StrCat("Failed to get length of the file '", file, "'")};
+    }
+    return size;
+  };
+
   if (it != meta.files.end()) {
+    meta.byte_size -= get_file_size(*it);
```
Comment (MBkkt) [inline core/index/index_writer.cpp:336]: add todo :)

**Pull request merged**